Add section to coding guidelines on Optional usage (#6894)
open-telemetry/opentelemetry-java-instrumentation: docs/contributing/style-guideline.md

  id("otel.nullaway-conventions")
}
```

## java.util.Optional usage

> `Optional` is not generally used in this project, following the reasoning from
> https://speakerdeck.com/trustin/writing-a-java-library-with-better-experience?slide=12.

Reviewer: ```suggestion
Following the reasoning from [Writing a Java library with better experience (slide 12)](https://speakerdeck.com/trustin/writing-a-java-library-with-better-experience?slide=12), 
usage of `java.util.Optional` is kept at a minimum in this project.
```

> It is ok to use `Optional` in places where it does not leak into public API signatures and where
> performance is not critical.

Reviewer: ```suggestion
Use `Optional` where it does not leak into public API signatures and where
performance is not a critical aspect.
```

Author: maybe `It is acceptable to use`? we don't exactly want to "recommend" it, only mention that it is "ok" (i.e. author's choice to use or not)
